### DISCUSSION TITLE
Can not use "go to definition" in Solargraph anymore

### DISCUSSION THREAD

**Discussion by pprotas** [LazyVim/LazyVim#990] [Q&A] [answered]
Solargraph is an LSP for Ruby. I know for certain that I used to be able to use "go to definition" `gd` to use Solargraph to go to the definition of the symbol under the cursor. But for some time now, this functionality has been broken in LazyVim. When I press `g` under a symbol, WhichKey doesn't even show "go to definition":
![image](https://github.com/LazyVim/LazyVim/assets/30273418/5bd013fc-06de-4228-afa9-1adcbe03c427)

But if I do it in, let's say, a TS file with `tsserver`, it works just as expected:
![image](https://github.com/LazyVim/LazyVim/assets/30273418/680b131c-9660-4111-aadc-d65b20cfa117)

LazyVim is ignoring it and not properly mapping the LSP action to a key. The reason I think that is because if I put my cursor over a symbol and type in `:lua vim.lsp.buf.definition()`, it works just as expected. So Solargraph is correctly communicating the "go to definition" capability to NeoVim.

Any clues on how to fix this?

**Comment by pprotas** [comment]
Though I'd update this, since I got it working.

**TLDR:**
Put this in your LSP configuration
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      setup = {
        solargraph = function()
          require("lazyvim.util").on_attach(function(client, _)
            if client.name == "solargraph" then
              client.server_capabilities.definitionProvider = true -- Force "definition" capability
            end
          end)
        end
      },
    },
  },
```

From my investigation I've seen that the "Go to definition" keybinding is only bound if the LSP attached to this buffer has the `definition` capability: https://github.com/LazyVim/LazyVim/blob/0b020dc37b30fd93a199f1124a95028cb544eac7/lua/lazyvim/plugins/lsp/keymaps.lua#L17

There is some other relevant code in the same file, for example the `has` method, that checks the LSP for capabilities.

I was quite confused by this, since Solargraph can definitely go to definition. But maybe it just doesn't register the capability in the way that LazyVim expects? After looking through  the output of `:= vim.lsp.get_active_clients()`, I found this in the Solargraph section:
```
16:01:57 msg_show   vim.lsp.get_active_clients()     dynamic_capabilities = {
16:01:57 msg_show   vim.lsp.get_active_clients()       capabilities = {
16:01:57 msg_show   vim.lsp.get_active_clients()         ["textDocument/definition"] = { {
16:01:57 msg_show   vim.lsp.get_active_clients()             id = "textDocument/definition",
16:01:57 msg_show   vim.lsp.get_active_clients()             method = "textDocument/definition",
16:01:57 msg_show   vim.lsp.get_active_clients()             registerOptions = {
16:01:57 msg_show   vim.lsp.get_active_clients()               definitionProvider = true <--- here is the dynamic_capability
16:01:57 msg_show   vim.lsp.get_active_clients()             }
16:01:57 msg_show   vim.lsp.get_active_clients()           } },
```
I am not familiar enough with `nvim-lsp` to say what this means, but it does look like Solargraph registers the "definition" capability as a **dynamic_capability**, instead of the regular way? I'll leave that up to smarter people than me to figure out.

In any case, after you apply the config in the way I described above, you will see this line in `:= vim.lsp.get_active_clients()`:
```
16:01:57 msg_show   vim.lsp.get_active_clients()     server_capabilities = {
16:01:57 msg_show   vim.lsp.get_active_clients()       completionProvider = {
16:01:57 msg_show   vim.lsp.get_active_clients()         resolveProvider = true,
16:01:57 msg_show   vim.lsp.get_active_clients()         triggerCharacters = { ".", ":", "@" }
16:01:57 msg_show   vim.lsp.get_active_clients()       },
16:01:57 msg_show   vim.lsp.get_active_clients()       definitionProvider = true, <-- Woohoo!

```

Which now means that LazyVim understands that Solargraph can provide definitions, and the keybind works as expected.

Hopefully this helps other people with other shitty LSPs that don't follow convention!

**Comment by folke** [comment]
What version of Neovim are you using?

**Reply by pprotas** [reply]
```
> nvim --version
NVIM v0.10.0-dev-3154+g8ea9a70d0-Homebrew
```

**Reply by folke** [reply]
that doesn't really tell me a lot. Is that a recent build? Like from the last couple of days?

**Reply by pprotas** [reply]
I'm following HEAD (`nightly` build) using Homebrew (I'm on macOS) https://github.com/neovim/neovim/wiki/Following-HEAD

I use the `brew upgrade neovim --fetch-HEAD` to upgrade my nvim. I think I ran it like a week ago, but this problem has been present for me for more than a month

**Reply by folke** [reply]
ok, thanks. Will check what's causing this. I actually did the work on dynamic registration in Neovim, so seems like something's wrong there (or on the LazyVim side)

**Reply by folke** [reply]
The keymap is properly created by LazyVim.
The issue is `Telescope`. It doesn't seem to properly resolve dynamic properties

**Comment by folke** [comment, marked as answer]
I just submitted a PR to Telescope to add dynamic registration capabilities.

See https://github.com/nvim-telescope/telescope.nvim/pull/2594
